### PR TITLE
VKD3D and VKD3D-Shader Interface cleanup

### PR DESCRIPTION
**Pull request by Joshua-Ashton**
Remove the next/type stuff from our interfaces as they aren't very useful at all and just seem to cause more issues than they are worth.

### CODE REVIEW

File: include/vkd3d.h [hunk 1]
```diff
@@ -108,16 +105,9 @@ struct vkd3d_instance_create_info
...
-/* Extends vkd3d_instance_create_info. Available since 1.1. */
-struct vkd3d_optional_instance_extensions_info
-{
-    enum vkd3d_structure_type type;
-    const void *next;
 
-    const char * const *extensions;
-    uint32_t extension_count;
+    const char * const *optional_instance_extensions;
+    uint32_t optional_instance_extension_count;
```
Comment (HansKristian-Work) [inline include/vkd3d.h:94]: SONAME breakage, need to update Meson build.

File: libs/vkd3d/device.c [hunk 2]
```diff
@@ -2039,23 +2024,31 @@ static HRESULT vkd3d_create_vk_device(struct d3d12_device *device,
...
-            &extension_count, &user_extension_supported)))
+            &extension_count, user_extension_supported))) {
+        vkd3d_free(user_extension_supported);
         return hr;
+    }
 
     vkd3d_physical_device_info_init(&device->device_info, device);
 
-    if (FAILED(hr = vkd3d_init_device_caps(device, create_info, &device->device_info)))
+    if (FAILED(hr = vkd3d_init_device_caps(device, create_info, &device->device_info))) {
```
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:2015]: Brace placement.

File: tests/d3d12_crosstest.h [hunk 3]
```diff
@@ -539,25 +539,25 @@ static HRESULT create_vkd3d_device(struct vkd3d_instance *instance,
...
-        .extension_count = ARRAY_SIZE(device_extensions),
-    };
 
     struct vkd3d_device_create_info device_create_info;
     VkPhysicalDevice vk_physical_device;
 
     if (!(vk_physical_device = select_physical_device(instance)))
         return E_INVALIDARG;
 
-    memset(&device_create_info, 0, sizeof(device_create_info));
```
Comment (HansKristian-Work) [inline tests/d3d12_crosstest.h:568]: Why not keep the memset?

File: include/vkd3d_shader.h [hunk 4]
```diff
@@ -214,6 +213,8 @@ struct vkd3d_shader_interface_info
     const struct vkd3d_shader_descriptor_binding *offset_buffer_binding;
 
     VkShaderStageFlagBits stage;
+
+    struct vkd3d_shader_transform_feedback_info *xfb_info;
```
Comment (HansKristian-Work) [inline include/vkd3d_shader.h:204]: `const struct` perhaps?

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Nits

File: tests/d3d12_crosstest.h (hunk 3 above)
Comment (Joshua-Ashton) [reply tests/d3d12_crosstest.h:568]: Oh, I was planning on refactoring the memsets to `(struct StructName) { }` initializers but I decided to do that elsewhere and not in this set these were left over, will fix.

File: libs/vkd3d/device.c (hunk 2 above)
Comment (Joshua-Ashton) [reply libs/vkd3d/device.c:2015]: Fixed.

File: tests/d3d12_crosstest.h (hunk 3 above)
Comment (Joshua-Ashton) [reply tests/d3d12_crosstest.h:568]: Fixed.

File: include/vkd3d_shader.h (hunk 4 above)
Comment (Joshua-Ashton) [reply include/vkd3d_shader.h:204]: Fixed.

File: libs/vkd3d/device.c [hunk 5]
```diff
@@ -2039,23 +2024,32 @@ static HRESULT vkd3d_create_vk_device(struct d3d12_device *device,
...
 
+    if (create_info->optional_device_extension_count)
+    {
+        if (!(user_extension_supported = vkd3d_calloc(create_info->optional_device_extension_count, sizeof(bool))))
+            return E_OUTOFMEMORY;
+    }
+
     if (FAILED(hr = vkd3d_init_device_extensions(device, create_info,
-            &extension_count, &user_extension_supported)))
+            &extension_count, user_extension_supported))) {
```
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:2008]: brace derp

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Small nit

**Review by HansKristian-Work — APPROVED** [review]
LGTM

**Comment by SveSop** [comment]
@HansKristian-Work 
Sadly this broke World of Warcraft (Retail) for me.
I was not able to make an apitrace while using wine, but i made one on my Win10 box (with GTX970 card), and that plays back just fine on the win 10 box.. however, when i try to play it back on linux with wine it crashes.

```
22 @0 IDXGIFactory6::EnumAdapterByGpuPreference(this = 0x2c928cc4000, Adapter = 1, GpuPreference
 = DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE, riid = IID_IDXGIAdapter1, ppvAdapter = &0x2c928cccad0)                                     
= S_OK
22: warning: failed with 0x887a0002 (DXGI_ERROR_NOT_FOUND): The item requested was not found. For GetPrivateData calls, this means t
hat the specified GUID had not been previously associated with the object.
24: debug: apitrace: warning: caught exception 0xc0000005
apitrace: warning: caught exception 0xc0000005
24: error: caught an unhandled exception
wine: Unhandled page fault on execute access to 0000000000B98D70 at address 0000000000B98D70 (thread 0024), starting debugger...
```
I used the binaries i found here: [https://github.com/Joshua-Ashton/apitrace/releases/tag/d3d12-alpha-2](https://github.com/Joshua-Ashton/apitrace/releases/tag/d3d12-alpha-2)

As i said, the Wow.trace file plays back fine on my windows box with d3dretrace.exe, but just crashes when i try that in wine.

You mentioned you have World of Warcraft installed? Could you check if it is me that is all messed up? :) (PS. f06f94bfb4ee0d367861753d228c7ce68d0c01e3 works like a champ tho).

**Comment by SveSop** [comment]
Fixed by 3d0913dc19d16a4f0d4cba9c6f4b807de269464c

**Pull request merged**